### DISCUSSION TITLE
Nested Layouts DX is pretty difficult

### DISCUSSION THREAD

**Discussion by Blankeos** [vikejs/vike#1681] [Discussion] [answered]
### Concerns
I did a bit more reading with [Nested Layouts](https://vike.dev/Layout#nested-layouts). The DX for layouts is kind of difficult :((. Is it possible to make it work a bit more like **SvelteKit** and **NextJS**?

Currently to do Nested Layouts:
- You don't use `+Layout` (because it overrides by default).
- You have to use `+route.tsx` to define routeParams?
- You have to use `+Page.tsx` that will act as a "Layout" and define the nested pages there as well?
- On top of this, to make it work, you have to have predefined strings and check the `pageContext.view` AND `<a href="..." keep-scroll-position />`.

I don't know, it just doesn't feel cohesive. Right now, I think this is one huge deal-breaker for me to use Vike :(

### Suggestions

I know Vike users don't prefer Next or other metaframeworks. I still think NextJS and SvelteKit nailed the DX for file-system routing really well. Here are some features that Vike currently lacks that can't replicate what you can do in Next:

1. **Should be able to make nested layouts without getting overridden:**

Currently by default, using `pages/route/+Layout` inside another page will override the `pages/+Layout` on the root. The override should probably be opt-in? Maybe a special file name for it like: `+LayoutOverride.tsx` while for the `+Layout`, it should work exactly like in NextJS or SvelteKit:

Example:
- `pages/+Layout.tsx` - Should be the root layout.
- `pages/my-route/+Layout.tsx` - Should be the nested layout inside `my-route`

    ```tsx
    // my-route's markup
    <RootLayout>
       <MyRouteLayout>
          {children}
       </MyRouteLayout>
    </RootLayout>
    ```

2. **Should be able to make folder structures using parenthesis so they're ignored as paths:**

This is useful for route grouping without affecting paths + grouping nested layouts without affecting paths.

When parenthesis is there:
- `pages/(horizontal)/settings/+Page.tsx` is equivalent to `/settings`
- `pages/(horizontal)/dashboard/+Page.tsx` is equivalent to `/dashboard`

Normally:
- `pages/horizontal/settings/+Page.tsx` is equivalent to `/horizontal/settings`
- `pages/horizontal/dashboard/+Page.tsx` is equivalent to `/horizontal/dashboard`

Let me know your thoughts!

**Comment by Blankeos** [comment]
In SvelteKit, this is how I usually address [this need (issue#1583)](https://github.com/vikejs/vike/issues/1583#issuecomment-2027980026) for "non-cumulative" layouts:

wherein (website) is the landing page and (app) is the dashboard.

<img width="265" alt="image" src="https://github.com/vikejs/vike/assets/38070918/0b3b0492-7048-4b0e-b6a8-0c5f62a0b8c5">

**Reply by brillout** [reply]
Regarding `(app)` and `(website)`, have you seen https://vike.dev/filesystem-routing#domain-driven-file-structure?

I agree that the succinctness of `()` is nice. Much better than https://vike.dev/filesystemRoutingRoot.

**Comment by brillout** [comment]
Quick spontaneous reaction: would making +Layout cumulative resolve your issues?

I'll think more about your feedback later.

**Reply by Blankeos** [reply]
Thanks for the quick reply as always @brillout! Yes it would. I'll also try to make this clear with some pictures and videos:

### Usecase why I need cumulative layouts. With the currently implementation, it's super difficult to this:

1. **Vertical Layout for the Landing Page**

- /pricing, /about, /faq, /login, /register pages that use that layout.
 
<img width="1028" alt="image" src="https://github.com/vikejs/vike/assets/38070918/a7fda076-0bea-4a90-9b7a-a0609674ce52">

2. **Horizontal Layout for the Dashboard Page**

- /local, /settings, /profile, /feed pages. (that don't have a /dashboard prefix)

<img width="1029" alt="image" src="https://github.com/vikejs/vike/assets/38070918/59f25512-39de-4d5b-b0d4-5bf8345095a8">

### In Vike, I would probably need to have a folder structure like this:

```sh
pages/
├── +Layout.tsx # RootLayout holds auth context.
├── landing/
│   ├── +Layout.tsx # HorizontalLayout has to rerun auth context here.
│   ├── about/+Page.tsx
│   ├── faq/+Page.tsx
│   ├── login/+Page.tsx
│   ├── register/+Page.tsx
│   └── pricing/+Page.tsx
└── app/
    ├── +Layout.tsx # VerticalLayout has to rerun auth context again here.
    ├── settings/+Page.tsx
    ├── profile/+Page.tsx
    └── feed/+Page.tsx
```

#### But a few tricky things with this:
- I'm forced to put `/landing/about` `/landing/faq` and `/landing/login` just for the "Landing Page" pages.
- I have to re-run the **Auth Context** for other layouts because I can only have 1 Layout for every page (which means there's no Root layout that can hold the important contexts that should be shared across all pages). - I did read about `Wrapper` but the docs don't really cover it that much and `vike-solid` doesn't seem to have it I think.
- If I have **shared state/signals that should persist between layout changes**, I can't do that either.

#### Here's a demonstration of this in the Vike + Solid app I'm working on:

https://github.com/vikejs/vike/assets/38070918/ff8d06b7-031b-4949-a969-61b184558e80

**Comment by Blankeos** [comment]
@brillout I tried porting everything to Solid Start to see if it makes a difference just so I can demonstrate it to you maybe.

https://github.com/vikejs/vike/assets/38070918/012054fc-9d84-4d1d-a8da-97e38f86bc17

Here are the codebases for both by the way:

- Vike: http://github.com/blankeos/solid-launch
- SolidStart: https://github.com/blankeos/solid-launch-start
- SvelteKit: https://github.com/blankeos/svelte-launch

**Comment by brillout** [comment]
Let me know if I'm missing something, but I think the conversation can be wrapped up in the following three feature requests:
- https://github.com/vikejs/vike/issues/1684
- https://github.com/vikejs/vike-react/issues/119
- https://github.com/vikejs/vike/issues/1685

**Reply by brillout** [reply, marked as answer]
Done: https://vike.dev/Layout#nested-layouts.

I went ahead and marked this reply as answer but feel free to disagree.
